### PR TITLE
feat(cheatcodes): Make `expectEmit` only work for the next call

### LINKED ISSUE #1745: bug: expectEmit and expectCall don't only check the next call
### Component

Forge

### Have you ensured that all of these are up to date?

- [ ] Foundry
- [ ] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (4fcd7e0 2022-05-26T00:08:37.693306Z)

### What command(s) is the bug in?

forge test

### Operating System

_No response_

### Describe the bug

My understanding is that the expected behavior of `expectEmit` is "look for event in the next call" but it seems to actually be "look for expected event before the end of the test". And the docs are contradictory here—one part says it must be the "next event" (which seems to imply it needs to be the first event in the next call), another part says "before the end of the current function". Similarly issue for `expectCall` behavior.

To ensure tests are clear, I think `expectEmit` and `expectCall` should only look in the next call. Otherwise this can result in bugs where your event was emitted _eventually_ in a later portion of your test, but not emitted at the time you expected it, and the test still passes. 

Reproduce with this test:

```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity 0.8.13;

import "forge-std/Test.sol";

contract CallMe {
  function calling() public {}
}

contract MyContract {
  CallMe callMe;

  constructor(CallMe _callMe) {
    callMe = _callMe;
  }

  event MyEvent();
  event OtherEvent();

  function noEventOrCall() public {}

  function myEventAndCall() public {
    emit OtherEvent(); // Shows event order does not matter if only checking for MyEvent.
    emit MyEvent();
    callMe.calling();
  }
}

contract ExpectTest is Test {
  event MyEvent();
  CallMe callMe;
  MyContract myContract;

  function setUp() public {
    callMe = new CallMe();
    myContract = new MyContract(callMe);
  }

  // -------- expectEmit Tests --------

  function test_MyEvent() public {
    // Passes like expected.
    vm.expectEmit(true, true, true, true);
    emit MyEvent();
    myContract.myEventAndCall();
  }

  function test_NoEventOrCall() public {
    // Fails like expected.
    vm.expectEmit(true, true, true, true);
    emit MyEvent();
    myContract.noEventOrCall();
  }

  function test_EventuallyMyEvent() public {
    // Passes, but should fail, because expectEmit should only check the next call.
    vm.expectEmit(true, true, true, true);
    emit MyEvent();
    myContract.noEventOrCall();
    myContract.myEventAndCall();
  }

  // -------- expectCall Tests --------

  function test_Call() public {
    // Passes like expected.
    vm.expectCall(address(callMe), abi.encodeWithSelector(CallMe.calling.selector));
    myContract.myEventAndCall();
  }

  function test_NoCall() public {
    // Fails like expected.
    vm.expectCall(address(callMe), abi.encodeWithSelector(CallMe.calling.selector));
    myContract.noEventOrCall();
  }

  function test_EventuallyCall() public {
    // Passes, but should fail, because expectCall should only check the next call.
    vm.expectCall(address(callMe), abi.encodeWithSelector(CallMe.calling.selector));
    emit MyEvent();
    myContract.noEventOrCall();
    myContract.myEventAndCall();
  }
}
```

### PR DESCRIPTION
**Pull request by Evalir**
*This is a big breaking change.*

Partially closes #1745. `expectCall` is left and needs #4912 to be merged before it can be reworked to support this. 

Makes `expectEmit` only work for the _immediate_ next call after the cheatcode has been called.

# Behavior

`expectEmit` remains _additive_. Calling `expectEmit` `N` times will set a lower bound of `N` events that must be seen for the cheatcode to pass successfully.

The current behavior of `expectEmit` allows to:
- declare the cheatcode before all events are emitted.
- "fill" the expected emits _before_ the test ends.
- match events after the tests ends.

The intended behavior is that it only works for the *next immediate call*. This new behavior only allows to:
- Call the cheatcode AND fill the expected emit right after calling the cheatcode. Declaring the cheatcode `N` times and then filling `N` events will fail.
- match events after the next CALL is triggered. Events not triggered by a call will be ignored. Events nested in the triggered CALL subtree will be matched. If the events defined are not matched, the execution will revert.

It's important to note the restriction imposed by this, namely that after the cheatcode is declared, no CALLs other than the one the cheatcode is intended to "match" must happen. This means that #1214 is now illegal behavior. Specifically, this means that this is now illegal:

```
function testEmitMyEvent_v2() external {
    cheatcodes.expectEmit(true, true, true, true);
    // myContract.myVar() is a call made after the cheatcode was called. This is illegal.
    // To fix this, move the call to before the cheatcode and use a variable.
    emit MyEvent(myContract.myVar(), address(this), address(this), block.timestamp);
    myContract.emitMyEvent();
}
````

## Why are now expect emits being pushed without subtracting 1 from the depth?
https://github.com/foundry-rs/foundry/pull/1215 is a great read to reason about depth. After this, the reason is that currently, `depth - 1` is being used is so that `expectEmit` checks the events after the call that invoked the cheatcode ends. However, we now wanna check the events after the next CALL (which will have more depth) ends, which means that we need to check for `depth`, not `depth - 1`.

When we're happy with this and merge it we should definitely prepare a telegram message & a tweetstorm for this change, as this might confuse people.

### CODE REVIEW

**Review by mattsse — APPROVED** [review]
Code-wise this looks good, but defer to @mds1 if this is the intended behavior now

**Comment by mds1** [comment]
> This means that https://github.com/foundry-rs/foundry/issues/1214 is now illegal behavior. Specifically, this means that this is now illegal:
> 
> ```solidity
> function testEmitMyEvent_v2() external {
>     cheatcodes.expectEmit(true, true, true, true);
>     // myContract.myVar() is a call made after the cheatcode was called. This is illegal.
>     // To fix this, move the call to before the cheatcode and use a variable.
>     emit MyEvent(myContract.myVar(), address(this), address(this), block.timestamp);
>     myContract.emitMyEvent();
> }
> ```

Is this change required if `myContract.myVar() is a staticcall? You cannot emit events in staticcalls / view methods, so perhaps they should be ignored automatically? Similar to how staticcalls are ignored when you're in a `startBroadcast` context

Aside from that this seems great and I agree with this behavior. This is a decent sizes breaking change though, and since we may have some other breaking changes, perhaps we should start feature flagging them? We could just check the value of `CARGO_PKG_VERSION_MAJOR` to disable certain features if that value is < 1, then delete the old code after v1 is released?

Will defer to you all on if we should do that or not. Either way we should maintain somewhere a list of the recent breaking changes we can continually point people to. Could be a changelog file, discussion in the repo, or a meta-issue

**Comment by Evalir** [comment]
Also, I believe this behavior will also break some forge-std tests (some seem to be failing), so we might need to follow up there

**Comment by Evalir** [comment]
@mds1 great catch on forge-std—i had missed this. With https://github.com/foundry-rs/foundry/pull/4920/commits/420d8146b69189450ae60196817bc9dded9dba92 we now have this behavior (will also update the main description):

A function emits events: `A, B, C, D, E, F, G`.

`expectEmit` will be able to match ranges with and without skipping events in between:
- `[A, B, C]` is valid.
- `[B, D, F]` is valid.
- `[G]` or any other single event combination is valid.
- `[B, A]` or similar out-of-order combinations are **invalid** (events must be in order).

File: evm/src/executor/inspector/cheatcodes/expect.rs
Comment (mattsse) [inline evm/src/executor/inspector/cheatcodes/expect.rs:136]: this is the same as `if state.expected_emits.iter().all(|expected| expected.found)`, right?

**Review by mattsse — APPROVED** [review]
smol nit, otherwise this looks sounds.

tests look good.

File: evm/src/executor/inspector/cheatcodes/expect.rs
Comment (Evalir) [reply evm/src/executor/inspector/cheatcodes/expect.rs:136]: yep indeed

Comment (Evalir) [reply evm/src/executor/inspector/cheatcodes/expect.rs]: done in https://github.com/foundry-rs/foundry/pull/4920/commits/d72a935f944a7a823e9c448dbdf39c6cbeb9151f , waiting for ci

**Comment by bowd** [comment]
Unless I'm missing something, there's one case that's impossible to fully cover with the given syntax:
Say we have a function on `ContractA` that calls `ContractB` which emits `EventB` and then `ContractA` emits `EventA`. 
We can't ensure that events originate from the contract we intend. I can do:
```
vm.expectEmit(true, true, true, true);
emit EventB(...);
emit EventA(...);
contractA.example();
```
But this would match irrespective of which contract emitted the events. This isn't allowed:
```
vm.expectEmit(true, true, true, true, address(contractB));
emit EventB(...);
vm.expectEmit(true, true, true, true, address(contractA));
emit EventA(...);
```

So essentially if I understand correctly, you can only match the contract address if you don't intend to match any other events emitted by other contracts. Feels like quite a big limitation. I'd suggest making it a bit more explicit in the docs.

**Comment by mds1** [comment]
@bowd can you open a separate issue so we can track this more easily? Your example seems like it should work to me so if it's currently not supported we can discuss in that issue with @Evalir

**Pull request merged**
